### ISSUE TITLE
RouteLink - Rendering based on RenderSettings.Mode

### ISSUE THREAD

**Issue by Mylan719** [riganti/dotvvm#781]
While investigating issue #779 I came across strange behavior of RoutLink control. If the ` RenderSettings.Mode` is set to `Server` it renders `a` tag with `href` attribute containing correctly formed url with query parameter values put in correct places.  Javascript logic needed to update the `href` attribute value on client side is also rendered:
```html
<a 
   href='/ControlSamples/RouteLink/RouteLinkQueryParameters?string=default&amp;int=5'
   onclick="event.stopPropagation();return !this.hasAttribute('disabled');" 
   class='link' 
   data-bind='attr: { &#39;href&#39;: &#39;/&#39; + "ControlSamples/RouteLink/RouteLinkQueryParameters" + dotvvm.buildUrlSuffix("", {"string": TestString,"int": TestInteger}) }, dotvvmEnable: true'>
   Test link
</a>
```

If the ` RenderSettings.Mode` is set to `Client` it renders the javascript logic, but there is no `href` attribute on the rendered `a` tag.
```html
<a 
   onclick="event.stopPropagation();return !this.hasAttribute('disabled');" 
   class='link' 
   data-bind='attr: { &#39;href&#39;: &#39;/&#39; + "ControlSamples/RouteLink/RouteLinkQueryParameters" + dotvvm.buildUrlSuffix("", {"string": TestString,"int": TestInteger}) }, dotvvmEnable: true'>
   Test link
</a>
```
If I understand philosophy of `RenderSettings.Mode` correctly it is so that `Client` means no server values are rendered in the original page source, just the javascript logic needed to update those values. Mode `Server` should render initial values from the viewmodel into the original page source as well as javascript logic for updating the values on the client side.

This is well and good in general but I find it strange to have `a` tag with no `href` attribute. Also not rendering links in the source is search engine unfriendly. And since `Client` mode is the default setting, this will result in poor discoverability of the web site and poor quality html for most of dotvvm users.

**Comment by tomasherceg** [comment]
If there is just a plain `RouteLink` in the HTML, we could definitely unify the output so we can hard-render the `href` and also provide the `data-bind` so it is updated whenever the values change.
But what shall we render when `RouteLink` is in a client-side rendered `Repeater`? There will be only one template item rendered and we don't have the values to render the route. Shall we just put there `href="#"` or something like that? I guess this won't help with the discoverability much.

I think this is a more fundamental issue - we should create some "hybrid-rendering" mode that would render everything on the server and then just "reanimate" the DOM elements with Knockout. This is how the server-rendering works in case of a plain `RouteLink` and other controls, but we don't have this solved for the `foreach` binding, and there are some limitations - you can never do this e.g. for the `IncludeInPage` property - if the initial value is false and can be changed to true later, you just need to render the stuff (even if it wouldn't be rendered in the server-rendering mode).

**Comment by exyi** [comment]
> we should create some "hybrid-rendering" mode

I think we have basically decided that this is what server-rendering should do. If you want server-only rendering, there is the resource binding. The problem is that we are not at all consistent in this.

As far as I understand what we want to achieve with Client/Server rendering, in this case it is right that controls don't prerender anything in Client rendering, but in some cases it seems better to make an exception and try to prerender whenever possible (the `Visible` property for example). RouteLink just seems to be a very good candidate for this prerendering since links are fundamental to the web.

**Comment by exyi** [comment]
This is already resolved by #1055
